File: 6/sentimental-readability/readability.py
from cs50 import get_string


def count_elements(text):
    letters = sum(c.isalpha() for c in text)

    words = len(text.split())

    sentences = text.count(".") + text.count("!") + text.count("?")

    return letters, words, sentences


text = get_string("Text: ")

letters, words, sentences = count_elements(text)


averageNumberOfLettersPer100Words = letters / words * 100

averageNumberOfSentencesPer100Words = sentences / words * 100

index = round(
    0.0588 * averageNumberOfLettersPer100Words
    - 0.296 * averageNumberOfSentencesPer100Words
    - 15.8
)

if index < 1:
    print("Before Grade 1")
elif index >= 16:
    print("Grade 16+")
else:
    print(f"Grade {index}")

File: 2_arrays/readability/readability.c
#include "cs50.h"
#include <ctype.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: ");

    // Count the number of letters, words, and sentences in the text
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Compute the Coleman-Liau index
    float L = ((float) letters / (float) words) * 100;
    float S = ((float) sentences / (float) words) * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;
    int grade = (int) (index + 0.5);
    if (grade >= 16)
    {
        // Print the grade level
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        // Print the grade level
        printf("Before Grade 1\n");
    }
    else
    {
        // Print the grade level
        printf("Grade %d\n", grade);
    }

    return 0;
}

int count_letters(string text)
{
    int letters = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

int count_words(string text)
{
    int words = 1;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isspace(text[i]))
        {
            words++;
        }
    }
    return words;
}

int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        char c = text[i];
        if (c == '!' || c == '?' || c == '.')
        {
            sentences++;
        }
    }
    return sentences;
}
